File: aoc15/ocaml/day_11/day_11.ml
let input = Input.chars "input" in

let next pwd =
  let rec next_rec pwd =
    match pwd with
    | [] -> []
    | 'z' :: rs -> 'a' :: next_rec rs
    | x :: rs -> Char.chr (Char.code x + 1) :: rs
  in
  List.rev (next_rec (List.rev pwd))
in

let rec straight pwd =
  match pwd with
  | [] -> false
  | a :: b :: c :: _
    when Char.code a = Char.code b - 1 && Char.code a = Char.code c - 2 ->
      true
  | _ :: ls -> straight ls
in
let invalid_letter pwd =
  List.mem 'i' pwd || List.mem 'o' pwd || List.mem 'l' pwd
in
let pairs pwd =
  let rec pairs_cnt pwd =
    match pwd with
    | [] -> 0
    | a :: b :: ls when a = b -> pairs_cnt ls + 1
    | _ :: ls -> pairs_cnt ls
  in
  pairs_cnt pwd >= 2
in
let valid pwd = straight pwd && (not (invalid_letter pwd)) && pairs pwd in

let next_valid pwd =
  let next_pwd = ref pwd in
  let quit = ref false in
  while not !quit do
    next_pwd := next !next_pwd;
    if valid !next_pwd then quit := true
  done;
  !next_pwd
in

let pwd = next_valid input in
let first = String.concat "" (List.map (String.make 1) pwd) in
Printf.printf "First: %s\n" first;

let pwd = next_valid pwd in
let second = String.concat "" (List.map (String.make 1) pwd) in
Printf.printf "Second: %s\n" second

File: aoc15/go/day_11/day_11.go
package main

import (
	"fmt"
	"os"
	"slices"
)

func next(pass []byte) {
	for i := len(pass) - 1; i >= 0; i-- {
		if pass[i] == 'z' {
			pass[i] = 'a'
		} else {
			pass[i]++
			break
		}
	}
}

func increasing(pass []byte) bool {
	for i := 2; i < len(pass); i++ {
		if pass[i] == pass[i-1]+1 && pass[i] == pass[i-2]+2 {
			return true
		}
	}
	return false
}

func invalid(pass []byte) bool {
	return slices.ContainsFunc(pass, func(c byte) bool {
		return c == 'i' || c == 'l' || c == 'o'
	})
}

func twopairs(pass []byte) bool {
	count := 0
	for i := 1; i < len(pass); i++ {
		if pass[i] == pass[i-1] {
			i++
			count++
		}
	}
	return count >= 2
}

func valid(pass []byte) bool {
	return increasing(pass) && !invalid(pass) && twopairs(pass)
}

func find(pass []byte) []byte {
	for {
		next(pass)
		if valid(pass) {
			return pass
		}
	}
}

func main() {
	input, _ := os.ReadFile("input")
	pass := input[:len(input)-1]

	fmt.Println("First:", string(find(pass)))
	fmt.Println("Second:", string(find(pass)))
}
